clang/test/SemaCUDA/target-overloads-in-function-prototypes.cu

```diff
@@ -0,0 +1,703 @@
+
+// (2.a) Prefer fitting overloads.
+template <typename T> TemplateTy template_vs_H_D_functions(T arg) { return {}; }
+__attribute__((device)) DeviceTy template_vs_H_D_functions(float arg) { return {}; }
+__attribute__((host)) HostTy template_vs_H_D_functions(float arg) { return {}; }
+
+__attribute__((device)) check<is_same<decltype(template_vs_H_D_functions(1.0f)), DeviceTy>::value>::type
+test_template_vs_H_D_functions_for_device() {
+  return TrueTy{};
+}
+
+__attribute__((host)) check<is_same<decltype(template_vs_H_D_functions(1.0f)), HostTy>::value>::type
+test_template_vs_H_D_functions_for_host() {
+  return TrueTy{};
+}
+
+__attribute__((host,device)) check<is_same<decltype(template_vs_H_D_functions(1.0f)), CurrentTargetTy>::value>::type
+test_template_vs_H_D_functions_for_hd() {
+  return TrueTy{};
+}
+
+// (2.b) Always prefer an HD candidate over a template candidate.
+template <typename T> TemplateTy template_vs_HD_function(T arg) { return {}; }
+__attribute__((host,device)) HostDeviceTy template_vs_HD_function(float arg) { return {}; }
+
+__attribute__((device)) check<is_same<decltype(template_vs_HD_function(1.0f)), HostDeviceTy>::value>::type
+test_template_vs_HD_function_for_device() {
+  return TrueTy{};
+}
+
+__attribute__((host)) check<is_same<decltype(template_vs_HD_function(1.0f)), HostDeviceTy>::value>::type
+test_template_vs_HD_function_for_host() {
+  return TrueTy{};
+}
+
+__attribute__((host,device)) check<is_same<decltype(template_vs_HD_function(1.0f)), HostDeviceTy>::value>::type
+test_template_vs_HD_function_for_hd() {
+  return TrueTy{};
+}
+
+
+// (2.c) For non-constexpr functions, prefer a sameside or native template
+// function over a wrongside non-template function:
+template <typename T> TemplateTy template_vs_D_function(T arg) { return {}; }
+__attribute__((device)) DeviceTy template_vs_D_function(float arg) { return {}; }
+
+__attribute__((host,device)) check<is_same<decltype(template_vs_D_function(1.0f)), TemplateIfHostTy>::value>::type
+test_template_vs_D_function_for_hd() {
+  return TrueTy{};
+}
+
+__attribute__((device)) check<is_same<decltype(template_vs_D_function(1.0f)), DeviceTy>::value>::type
+test_template_vs_D_function_for_device() {
+  return TrueTy{};
+}
+
+__attribute__((host)) check<is_same<decltype(template_vs_D_function(1.0f)), TemplateTy>::value>::type
+test_template_vs_D_function_for_host() {
+  return TrueTy{};
+}
+
+
+// If only a wrongside function is available, it is selected.
+__attribute__((device)) DeviceTy only_D_function(float arg) { return {}; }
+
+__attribute__((host)) check<is_same<decltype(only_D_function(1.0f)), DeviceTy>::value>::type
+test_only_D_function_for_host() {
+  return TrueTy{};
+}
+
+// Default arguments for template parameters occur before the target attribute,
+// so we can't identify the "right" overload for them.
+template <typename T = targetdep_t>
+__attribute__((device)) // expected-warning {{target attribute has been ignored for overload resolution}}
```

Comment:
A point worth noting is this case, where moving the target attribute before the overloaded call (here wrapped in `targetdep_t`) is not possible: in default arguments for template parameters. The rocPRIM library contains this kind of code (and therefore also the introduced warning). I'd be happy about suggestions to improve this.

Reply:
We could mitigate this issue at least for non-type template parameters with default arguments by supporting this syntax to control overload resolution:
```CUDA
template <__attribute__((device)) int V = device_specific_init()>
__attribute__((device)) void foo(void) {
  // ...
}
```
Somewhat surprisingly (to me), trunk already accepts device attributes in non-type template arguments without diagnostic (but seems to ignore them), so this would not introduce new syntax.
So far, I have not encountered type template parameters with target-dependent default arguments -- where this syntax is not possible -- outside of this test. The problematic cases in rocPRIM are non-type template parameters (e.g. the device function `device_warp_size` in the default argument [here](https://github.com/ROCm/rocPRIM/blob/100b3ea65b61edf5af0b7f3fb98b8670cce51c2a/rocprim/include/rocprim/block/block_load_func.hpp#L372C29-L372C45)).

@yxsamliu, do you think that this is a viable way to avoid the warnings, e.g., in rocPRIM?

Reply:
@yxsamliu, I opened #109663 with an alternative solution for the root problem that I'm actually trying to solve (diagnosing wrong uses of `__AMDGCN_WAVEFRONT_SIZE`), which would make PRs #91478, #93546, and #103031 (this one) unnecessary.
Please let me know if you think we should rather pursue #109663.